File: 5. JavaScript (node.js)/tema1.js
// Importar módulos	
const axios = require('axios');
const cheerio = require('cheerio');
const fs = require('fs');
const { Canvas } = require('skia-canvas');
const WordCloud = require('node-wordcloud')((w, h) => new Canvas(w, h));

// Declaración de constantes
const URL = 'https://cienciasdelsur.com/'
const OUTPUT_FILE_ARTICLES_CONTENT = 'EXTRACCION_TEXTOS.txt'
const OUTPUT_FILE_FREQUENCIES = 'FRECUENCIAS_PALABRAS.txt'
const WORDS_TO_DISMISS = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'p', 'q',
    'r', 's', 't', 'u', 'v', 'y', 'x', 'w', 'z', 'y', 'o', 'no', 'el', 'la', 'los', 'las',
    'un', 'una', 'es', 'su', 'unos', 'unas', 'pero', 'a', 'ante', 'bajo', 'cabe', 'con', 'contra', 'de',
    'desde', 'durante', 'en', 'entre', 'hacia', 'hasta', 'mediante', 'para', 'por', 'según',
    'sin', 'so', 'sobre', 'tras', 'como', 'que', 'se', 'del', 'al', 'lo', 'como', 'esta',
    'está', 'este', 'estos', 'estas', 'ha', 'han', 'porque', 'si', 'sí', 'sino', 'más', 'menos',
    'ser', 'estar', 'fue', 'son', 'sin', 'tal', 'también', 'todo', 'todos', 'cual', 'cuales', 'cuya', 'cuyos'
]

// Obtenemos los divs con la clase 'td-module-thumb' y luego recorremos para extraer los enlaces de los articulos
const get_links_articles = async () => {
    try {
        let aux_links_articles = [];
        const response = await axios.get(URL);
        const $ = cheerio.load(response.data);
        $('div.td-module-thumb').each((i, el) => {
            const tag_a = $(el).find('a');
            const link = tag_a.attr('href');
            const title = tag_a.attr('title');
            aux_links_articles.push({
                link,
                title
            });
        });
        // Eliminamos los enlaces duplicados
        const links_articles_unique = aux_links_articles.filter((link, index, self) =>
            index === self.findIndex((t) => (
                t.link === link.link
            ))
        );
        return links_articles_unique;
    } catch (error) {
        console.error(error);
    }
}

// Obtenemos el contenido de los articulos
const get_articles_content = async (links_articles) => {
    try {
        let aux_articles_content = [];
        for (let i = 0; i < links_articles.length; i++) {
            console.log(`⭕ Iniciando scraping en el articulo: ${links_articles[i].title}`);
            const response = await axios.get(links_articles[i].link);
            const $ = cheerio.load(response.data);
            const article_content = $('.td-post-content');
            // Extraemos el texto de todos los parrafos
            let aux_text = '';
            article_content.find('p').each((i, el) => {
                // Extraer solamente el texto e ignorar las etiquetas de imagen
                if ($(el).find('img').length === 0) {
                    aux_text += $(el).text() + '\n';
                }
            });
            aux_articles_content.push(aux_text);
            response ? console.log(`✅ Finalizado scraping en el articulo: ${links_articles[i].title}\n`) : console.log(`❌ Error al realizar scraping en el articulo: ${links_articles[i].title}\n`);
        }

        return aux_articles_content;
    } catch (error) {
        console.error(error);
    }
}

const main = async () => {
    // Se obtienen los enlaces de los articulos y su titulo
    const links_articles = (await Promise.all([get_links_articles()]))[0];

    // Se extrae el contenido de los articulos
    const articles_content = (await Promise.all([get_articles_content(links_articles)]))[0];

    // Escribimos el contenido de los articulos en un archivo de texto
    try {
        fs.writeFileSync(OUTPUT_FILE_ARTICLES_CONTENT, articles_content.join('\n'));
        console.log(`📝 Archivo ${OUTPUT_FILE_ARTICLES_CONTENT} creado correctamente`);
    } catch (error) {
        console.error(error);
    }

    console.log(`\n-------------------------------------------------------------------------------`);

    // Leemos el archivo de texto con el contenido de los articulos
    console.log('\n📊 Iniciando análisis de frecuencias de palabras...')
    const data = fs.readFileSync(OUTPUT_FILE_ARTICLES_CONTENT, 'utf8');

    // Se convierte el texto a minusculas, se eliminan los caracteres especiales y se separa por palabras
    const words = data.toLowerCase().replace(/[^a-zñáéíóúü ]/g, "").split(/\s+/);

    // Se eliminan las palabras que no se desean tener en cuenta
    const words_filtered = words.filter(word => !WORDS_TO_DISMISS.includes(word));

    // Se crea un objeto con las palabras y sus frecuencias
    const words_frequencies = words_filtered.reduce((acc, word) => {
        acc[word] = acc[word] ? acc[word] + 1 : 1;
        return acc;
    }, {});

    // Se ordenan las palabras por frecuencia y se exporta a un archivo
    const words_frequencies_sorted = Object.entries(words_frequencies).sort((a, b) => b[1] - a[1]);
    fs.writeFileSync(OUTPUT_FILE_FREQUENCIES, words_frequencies_sorted.map(word => word.join(' - ')).join('\n'));
    console.log(`📝 Se ha generado el archivo: ${OUTPUT_FILE_FREQUENCIES} con las frecuencias de las palabras!`)

    // Mostramos las 50 palabras más frecuentes
    console.log('--> 📊 Las 50 palabras más frecuentes son:')
    console.log('\t', words_frequencies_sorted.slice(0, 50).map(word => word.join(' - ')).join('\n\t'))

    console.log(`\n-------------------------------------------------------------------------------`);
    
    // Se genera la nube de palabras de las 50 palabras más frecuentes
    console.log('\n ☁ Generando nube de palabras...')
    const list = words_frequencies_sorted.slice(0, 50).map(word => ([word[0], word[1]]));

    const colorPanel = ['#54b399', '#6092c0', '#d36086', '#9170b8', '#ca8eae', '#d6bf57', '#b9a888', '#da8b45', '#aa6556', '#e7664c']

    const options = {
        gridSize: 8,
        rotateRatio: 1,
        rotationSteps: 7,
        rotationRange: [-70, 70],
        backgroundColor: '#fff',
        sizeRange: [18, 70],
        color: function (word, weight) {
            return colorPanel[Math.floor(Math.random() * colorPanel.length)]
        },
        fontWeight: 'bold',
        fontFamily: `"PingFang SC", "Microsoft YaHei", "Segoe UI Emoji", "Segoe UI Emoji","Segoe UI Historic"`,
        shape: 'square'
    }

    const canvas = new Canvas(500, 500);
    const wordcloud = WordCloud(canvas, { list, ...options })
    wordcloud.draw()
    canvas.toBuffer().then(buffer => {
        fs.writeFileSync('nube_de_palabras.png', buffer)
        console.log('✅ Nube de palabras generada correctamente!')
        console.log('🎉 Fin del programa!')
    })
}


main();

File: 4. C#/tema1/tema1.cs
using System;
using System.Collections.Generic;
using System.Linq;
using HtmlAgilityPack;
using ScrapySharp.Extensions;
using ScrapySharp.Network;
using System.IO;
using System.Net;
using System.Text.RegularExpressions;
using KnowledgePicker.WordCloud;
using KnowledgePicker.WordCloud.Coloring;
using KnowledgePicker.WordCloud.Drawing;
using KnowledgePicker.WordCloud.Layouts;
using KnowledgePicker.WordCloud.Primitives;
using KnowledgePicker.WordCloud.Sizers;
using SkiaSharp;

namespace WebScraper {
    // Declaracion de constantes
    public static class Constants {
        public const string URL = "https://cienciasdelsur.com/";
        public const string OUTPUT_FILE_ARTICLES_CONTENT = "EXTRACCION_TEXTOS.txt";
        public const string OUTPUT_FILE_FREQUENCIES = "FRECUENCIAS_PALABRAS.txt";
        public static readonly List<string> WORDS_TO_DISMISS = new List<string> {
            "a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "p", "q", 
            "r", "s", "t", "u", "v", "y", "x", "w", "z", "y", "o", "no", "el", "la", "los", "las", 
            "un", "una", "es", "su", "unos", "unas" , "pero", "a", "ante", "bajo", "cabe", "con", "contra", "de", 
            "desde", "durante", "en", "entre", "hacia", "hasta", "mediante", "para", "por", "según", 
            "sin", "so", "sobre", "tras", "como", "que", "se", "del", "al", "lo", "como", "esta", 
            "está", "este", "estos", "estas", "ha", "han", "porque", "si", "sí", "sino", "más", "menos", 
            "ser", "estar", "fue", "son", "sin", "tal", "también", "todo", "todos", "cual", "cuales", "cuya", "cuyos"
        };
    }

    class Program {
        static void Main(string[] args) {
            // Creamos la instancia de la clase ScrapingBrowser
            var browser = new ScrapingBrowser();

            // Hacemos la petición y obtenemos el objeto WebPage
            WebPage webpage = browser.NavigateToPage(new Uri(Constants.URL));
            HtmlNode html = webpage.Html;

            // Obtenemos todos los div que tengan la clase "td-module-thumb" (contenedor de los artículos)
            var articles_container = html.CssSelect("div.td-module-thumb");

            // Obtenemos el link de cada articulo y lo guardamos en una lista
            var links_articles = new List<Dictionary<string, string>>();
            var links_aux = new List<string>();
            foreach (var article in articles_container) {
                string title = article.CssSelect("a").Single().GetAttributeValue("title", "");
                string link = article.CssSelect("a").Single().GetAttributeValue("href", "");

                // Si el link no está en la lista, lo agregamos
                if (!links_aux.Contains(link)) {
                    var dict = new Dictionary<string, string>();
                    dict.Add("title", title);
                    dict.Add("link", link);
                    links_articles.Add(dict);
                    links_aux.Add(link);
                }
            }

            // Iteramos sobre la lista de links y sacamos el contenido de cada uno
            var all_text_articles = new List<string>();
            foreach (var link in links_articles) {
                Console.WriteLine("* Iniciando scraping en el articulo: " + link["title"] + "...");
                var webpage2 = browser.NavigateToPage(new Uri(link["link"]));
                var article_content = webpage2.Html.CssSelect("div.td-post-content").Single();
                var article_text_content = article_content.CssSelect("p");
                foreach (var paragraph in article_text_content) {
                    all_text_articles.Add(paragraph.InnerText);
                }

                Console.WriteLine($"- Se finalizo el scraping en el articulo: {link["title"]}\n");
            }

            // Guardamos el contenido de todos los articulos en un archivo de texto
            using (StreamWriter file = new StreamWriter(Constants.OUTPUT_FILE_ARTICLES_CONTENT)) {
                foreach (var text in all_text_articles) {
                    file.WriteLine(text + " ");
                }
            }

            Console.WriteLine($"** Se ha generado el archivo: {Constants.OUTPUT_FILE_ARTICLES_CONTENT} con el contenido de todos los articulos scrapeados!");
            Console.WriteLine("-----------------------------------------------------------------------------------------");
            // Leemos el archivo de texto generado utilizando la codificacion UTF-8
            Console.WriteLine("**  Iniciando análisis de frecuencias de palabras...");
            string text_file = File.ReadAllText(Constants.OUTPUT_FILE_ARTICLES_CONTENT, System.Text.Encoding.UTF8);

            // Convertimos todo a minúsculas, eliminamos los caracteres especiales y separamos las palabras
            text_file = text_file.ToLower();
            text_file = Regex.Replace(text_file, @"[^a-zñáéíóúü\s]", "");
            var palabras = text_file.Split(' ');

            // Eliminamos todos los espacios en blanco
            palabras = palabras.Where(palabra => !string.IsNullOrWhiteSpace(palabra))
                            .Select(palabra => palabra.Trim())
                            .ToArray();

            // Contamos las frecuencias de cada palabra, sin contar las palabras que no nos interesan
            Dictionary<string, int> frecuencias = new Dictionary<string, int>();
            foreach (var palabra in palabras) {
                if (!(Constants.WORDS_TO_DISMISS.Contains(palabra))) {
                    if (frecuencias.ContainsKey(palabra)) {
                        frecuencias[palabra]++;
                    } else {
                        frecuencias[palabra] = 1;
                    }
                }
            }

            // Ordenamos las frecuencias de mayor a menor
            var frecuencias_ordenadas = frecuencias.OrderByDescending(x => x.Value);

            // Guardamos las frecuencias en un archivo de texto
            using (StreamWriter file = new StreamWriter(Constants.OUTPUT_FILE_FREQUENCIES, false, System.Text.Encoding.UTF8)) {
                foreach (var item in frecuencias_ordenadas) {
                    file.WriteLine($"{item.Key}:{item.Value}");
                }
            }

            // Mostramos las 50 palabras más frecuentes
            Dictionary<string, int> frecuencias_ordenadas_50 = new Dictionary<string, int>();
            int contador = 0;
            foreach (var item in frecuencias_ordenadas) {
                if (contador == 50) {
                    break;
                }
                frecuencias_ordenadas_50[item.Key] = item.Value;
                contador++;
            }
            Console.WriteLine($"** Las 50 palabras más frecuentes son:");
            foreach (var item in frecuencias_ordenadas_50) {
                Console.WriteLine($"\t--> {item.Key}:{item.Value}");
            }

            Console.WriteLine($"** Se ha generado el archivo: {Constants.OUTPUT_FILE_FREQUENCIES} con las frecuencias de palabras!");
            Console.WriteLine("-----------------------------------------------------------------------------------------");

            // Generar la nube
            Console.WriteLine("** Generando nube de palabras...");
            
            // Configuración de la nube de palabras.
            const int k = 6; // scale
            var wordCloud = new WordCloudInput(frecuencias_ordenadas_50.Select(p => new WordCloudEntry(p.Key, p.Value))) {
                Width = 1024 * k,
                Height = 256 * k,
                MinFontSize = 16 * k,
                MaxFontSize = 32 * k
            };

            var sizer = new LogSizer(wordCloud);
            using var engine = new SkGraphicEngine(sizer, wordCloud);
            var layout = new SpiralLayout(wordCloud);
            var colorizer = new RandomColorizer();
            var wcg = new WordCloudGenerator<SKBitmap>(wordCloud, engine, layout, colorizer);

            // Dibuja la nube de palabras.
            using var final = new SKBitmap(wordCloud.Width, wordCloud.Height);
            using var canvas = new SKCanvas(final);
            canvas.Clear(SKColors.White);
            using var bitmap = wcg.Draw();
            canvas.DrawBitmap(bitmap, 0, 0);

            // Guardar la nube de palabras.
            using var data = final.Encode(SKEncodedImageFormat.Png, 100);
            using var stream = File.OpenWrite("NUBE_PALABRAS.png");
            data.SaveTo(stream);
            Console.WriteLine($"** Se ha generado el archivo: NUBE_PALABRAS.png con la nube de palabras!");
        }
    }
}